lib/set.js

```diff
@@ -59,13 +67,12 @@ Set.prototype.addMigration = function(title, up, down){
  */
 
 Set.prototype.save = function(fn){
-  var self = this
-    , json = JSON.stringify(this);
-  fs.writeFile(this.path, json, function(err){
-    if (err) return fn(err);
-
+  var self = this;
+  this.state.save({
```

Comment:
Of course! It's easier to fix that before allowing people to come with their own implementations.

When that is said, I think that it is impossible to solve this problem in a general way. Consider this case:

Developer A and B both branch out of master with feature branches called feature/a and feature/b respectively. They both add a migration. Developer A merges his branch first. From there, we have two possible cases:

1) Developer B proceeds to merge his branch to master, and run the migrations. The state in the database has all migrations from master, and the one from feature/b already. There's no way of knowing if the result of applying the migration from feature/a on top of the database with master + feature/b is the same result as applying what's on master + feature/a + feature/b.

2) The other option is that Developer B goes on to rebase his branch on top of the new master with feature/a on it already, and make sure that his migration applies cleanly on top of that before merging it.

Option 2 is the only viable one, as there is no way to know if the result of doing those two migrations will be the same. If there is a way, it's going to be specific to database implementations and thus out of scope for this module - and even then, I'd be very cautious of actually relying on something like that.

Thus, the only options we have revolve around forcing/pushing people into choosing option 2. But no matter what we do (expect from forcing people into having non-descriptive migration names - e.g. 001.js, 002.js ...) it will ultimately be based on convention.

If you cannot trust that people are disciplined about using migrations as append-only lists, I really don't think that we can come up with a generic solution.